client/src/components/utils/SASCodeDocument.ts

```diff
@@ -0,0 +1,249 @@
+    const firstLineNumber = lineNumberInWrappedCode[0];
+    const lastLineNumber = lineNumberInWrappedCode[lastIndex];
+
+    // if the problem occurs before the first raw code line,
+    // then re-locate it at the first character in the first raw code line.
+    if (lineNumberInLog < firstLineNumber) {
+      return {
+        lineNumber: firstLineNumber + offsets[0].lineOffset,
+        startColumn: 0,
+        endColumn: 1,
+      };
+    }
+
+    // if the problem occurs after the last raw code line,
+    // then re-located it at the last character in the last raw code line.
+    if (lineNumberInLog > lastLineNumber) {
+      const codeList = this.getRawCode().split("\n");
+      const count = codeList[codeList.length - 1].length;
+      const lastCharacterIndex = count === 0 ? 0 : count - 1;
+      return {
+        lineNumber: lastLineNumber + offsets[lastIndex].lineOffset,
+        startColumn: lastCharacterIndex,
+        endColumn: lastCharacterIndex + 1,
+      };
+    }
+  }
+
+  private codeIsEmpty(code: string): boolean {
+    return code.trim() === "";
+  }
+
+  private wrapCodeWithSASProgramFileName(
+    code: string,
+    fileName: string | undefined,
+  ): string {
+    if (fileName === undefined) {
+      return code;
+    } else {
+      fileName = fileName.replace(/[('")]/g, "%$&");
+      const wrapped =
+        "%let _SASPROGRAMFILE = %nrquote(%nrstr(" + fileName + "));\n" + code;
+      return wrapped;
+    }
+  }
+
+  private wrapCodeWithPreambleAndPostamble(
+    code: string,
+    preamble?: string,
+    postamble?: string,
+  ): string {
+    return (
+      (preamble ? preamble + "\n" : "") +
+      code +
+      (postamble ? "\n" + postamble : "")
+    );
+  }
+
+  private wrapCodeWithOutputHtml(code: string): string {
+    if (this.options.outputHtml) {
+      const htmlStyle = this.options.htmlStyle.trim();
+      const htmlStyleOption = htmlStyle !== "" ? ` style=${htmlStyle}` : "";
+      const outputDestination = this.options?.uuid
+        ? ` body="${this.options.uuid}.htm"`
+        : "";
+
+      return `title;footnote;ods _all_ close;
+ods graphics on;
+ods html5(id=vscode)${htmlStyleOption} options(bitmap_mode='inline' svg_mode='inline')${outputDestination};
+${code}
+;*';*";*/;run;quit;ods html5(id=vscode) close;`;
+    } else {
+      return code;
+    }
+  }
+
+  private wrapSQL(code: string) {
+    return `proc sql;
+${code}
+;quit;`;
+  }
+
+  private wrapPython(code: string) {
+    return `proc python;
+submit;
+${code}
+endsubmit;
+run;`;
+  }
+
+  private insertLogStartIndicator(code: string): string {
+    // add a comment line at the top of code,
+    // this comment line will be used as indicator to the beginning of log related with this code
+    return `/** LOG_START_INDICATOR **/
+${code}`;
+  }
+
+  private fillBlankLines(code: string): string {
+    const filledCode = code
+      .split("\n")
+      .map((line) => (line.trim() === "" ? `/* BLANK LINE*/` : line))
```

Comment:
this may break user's code. we should not do this

Reply:
good point, working on it